src/util/mp_arith.h

```diff
@@ -105,28 +114,53 @@ struct numeric_castt<T,
 ///         empty optional otherwise.
 template <typename T>
 struct numeric_castt<T,
+                     mp_integer,
                      typename std::enable_if<std::is_integral<T>::value &&
                                              !std::is_signed<T>::value>::type>
 {
   static optionalt<T> numeric_cast(const mp_integer &mpi)
   {
+#if !defined(_MSC_VER) || _MSC_VER >= 1900
     static_assert(
       std::numeric_limits<T>::max() <=
           std::numeric_limits<decltype(mpi.to_ulong())>::max() &&
         std::numeric_limits<T>::min() >=
           std::numeric_limits<decltype(mpi.to_ulong())>::min(),
       "Numeric cast only works for types smaller than unsigned long long");
+#else
+    // std::numeric_limits<> methods are not declared constexpr in old versions
+    // of VS
+    PRECONDITION(
+      std::numeric_limits<T>::max() <=
+        std::numeric_limits<decltype(mpi.to_ulong())>::max() &&
+      std::numeric_limits<T>::min() >=
+        std::numeric_limits<decltype(mpi.to_ulong())>::min());
+#endif
     if(mpi <= std::numeric_limits<T>::max() && mpi >= 0)
-      return {static_cast<T>(mpi.to_ulong())};
+      return static_cast<T>(mpi.to_ulong());
     else
       return {};
   }
 };
 
-template <typename T>
-optionalt<T> numeric_cast(const mp_integer &mpi)
+template <typename Target, typename Source>
+optionalt<Target> numeric_cast(const Source &arg)
+{
+  return numeric_castt<Target, Source>::numeric_cast(arg);
+}
+
+/// An invariant with fail with message "Bad conversion" if conversion
```

Comment:
s/with/will/? Also, the first sentence should say what it does, not just when it fails.